MCP USE shows Agent stopped after reaching Max Interations

**Describe the bug**
I gave the max steps limit as 100 and ran my code using playwright mcp tools using gemini llm with langchain as provider. As you see this log step 3/100 but max interaction for execeded 100. how ??

2025-08-06 21:19:31,417 - mcp_use - INFO - 👣 Step 3/100
2025-08-06 21:19:31 - mcp_use - INFO - 👣 Step 3/100
2025-08-06 21:19:33,413 - mcp_use - INFO - ✅ Agent finished at step 3
2025-08-06 21:19:33 - mcp_use - INFO - ✅ Agent finished at step 3
2025-08-06 21:19:33,413 - mcp_use - WARNING - ⚠️ Agent stopped after reaching max iterations (100)
2025-08-06 21:19:33 - mcp_use - WARNING - ⚠️ Agent stopped after reaching max iterations (100)


**To Reproduce**
Steps to reproduce the behavior:
I dont know but this is a very frequent problem for i used lanchain based gemini combined with mcp_use to make mcp_agent and use how testing automation with playwright mcp

**Expected behavior**
Steps are still pending so i dont know error must not occur and interrupt my code in between.




**Desktop (please complete the following information):**
 - OS: mac os using mcp-use = "^1.3.6"

2025-08-12 11:08:59,682 - mcp_use - INFO - 🔧 Tool call: browser_click with input: {'ref': 'e15', 'element': 'button "Login"'}
2025-08-12 11:08:59 - mcp_use - INFO - 🔧 Tool call: browser_click with input: {'ref': 'e15', 'element': 'button "Login"'}
2025-08-12 11:08:59,682 - mcp_use - INFO - 📄 Tool result: ### Ran Playwright code ```js await page.locator('[data-test="login-button"]').click(); ```  ### ...
2025-08-12 11:08:59 - mcp_use - INFO - 📄 Tool result: ### Ran Playwright code ```js await page.locator('[data-test="login-button"]').click(); ```  ### ...
2025-08-12 11:08:59,682 - mcp_use - INFO - 👣 Step 7/1000
2025-08-12 11:08:59 - mcp_use - INFO - 👣 Step 7/1000
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence] Entering Chain run with input:
{
  "input": ""
}
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad>] Entering Chain run with input:
{
  "input": ""
}
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad> > chain:RunnableParallel<agent_scratchpad>] Entering Chain run with input:
{
  "input": ""
}
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad> > chain:RunnableParallel<agent_scratchpad> > chain:RunnableLambda] Entering Chain run with input:
{
  "input": ""
}
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad> > chain:RunnableParallel<agent_scratchpad> > chain:RunnableLambda] [2ms] Exiting Chain run with output:
[outputs]
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad> > chain:RunnableParallel<agent_scratchpad>] [4ms] Exiting Chain run with output:
[outputs]
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence > chain:RunnableAssign<agent_scratchpad>] [6ms] Exiting Chain run with output:
[outputs]
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence > prompt:ChatPromptTemplate] Entering Prompt run with input:
[inputs]
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence > prompt:ChatPromptTemplate] [1ms] Exiting Prompt run with output:
[outputs]
[llm/start] [tool:user_actions_enhancer > chain:RunnableSequence > llm:ChatGoogleGenerativeAI] Entering LLM run with input:
{
  "prompts": [
    "System: You are a helpful AI assistant.\nYou have access to the following tools:\n\n- browser_close: Close the page\n- browser_resize: Resize the browser window\n- browser_console_messages: Returns all console messages\n- browser_handle_dialog: Handle a dialog\n- browser_evaluate: Evaluate JavaScript expression on page or element\n- browser_file_upload: Upload one or multiple files\n- browser_install: Install the browser specified in the config. Call this if you get an error about the browser not being installed.\n- browser_press_key: Press a key on the keyboard\n- browser_type: Type text into editable element\n- browser_navigate: Navigate to a URL\n- browser_navigate_back: Go back to the previous page\n- browser_navigate_forward: Go forward to the next page\n- browser_network_requests: Returns all network requests since loading the page\n- browser_take_screenshot: Take a screenshot of the current page. You can't perform actions based on the screenshot, use browser_snapshot for actions.\n- browser_snapshot: Capture accessibility snapshot of the current page, this is better than screenshot\n- browser_click: Perform click on a web page\n- browser_drag: Perform drag and drop between two elements\n- browser_hover: Hover over element on page\n- browser_select_option: Select an option in a dropdown\n- browser_tab_list: List browser tabs\n- browser_tab_new: Open a new tab\n- browser_tab_select: Select a tab by index\n- browser_tab_close: Close a tab\n- browser_wait_for: Wait for text to appear or disappear or a specified time to pass\n\nUse the following format:\n\nQuestion: the input question you must answer\nThought: you should always think about what to do\nAction: the action to take, should be one of the available tools\nAction Input: the input to the action\nObservation: the result of the action\n... (this Thought/Action/Action Input/Observation can repeat N times)\nThought: I now know the final answer\nFinal Answer: the final answer to the original input question\nHuman: \n**TASK**: Generate enhanced user_actions.json with comprehensive selector strategies and testing capabilities.\n\n**CRITICAL REQUIREMENTS**:\n- Take parsed_actions.json as base structure\n- Enhance EACH action with comprehensive selector strategies\n- Add wait conditions and validation for timing-sensitive actions\n- EXECUTE THE COMPLETE SCRIPT to validate selector accuracy\n\n**SELECTOR PRIORITY ORDER** (Use this exact order for ALL selector arrays):\n1. data-test attributes (most reliable)\n2. accessibility selectors ([aria-label], [role], [aria-*])\n3. ID selectors (#element-id)\n4. class selectors (.class-name)\n5. xpath expressions (//element[@attribute='value'])\n6. text-based selectors (element:has-text('Text'))\n\n**ENHANCED ACTION SCHEMA**:\n```json\n{\n  \"action_id\": 1,\n  \"action_type\": \"click\",\n  \"primary_selector\": \"[data-test='username']\",\n  \"backup_selectors\": [\n    \"[aria-label='Username']\",\n    \"#username-field\",\n    \".username-input\",\n    \"//input[@data-test='username']\",\n    \"input[placeholder='Username']\"\n  ],\n  \"value\": \"user@bank.com\",\n  \"description\": \"Fill the username input field\",\n  \"wait_conditions\": {\n    \"wait_type\": \"element_clickable\",\n    \"timeout_ms\": 5000,\n    \"wait_selectors\": [\n      \"[data-test='username']\",\n      \"[aria-label='Username']\",\n      \"#username-field\"\n    ]\n  },\n  \"validation_after\": {\n    \"expected_state\": \"field_filled\",\n    \"validation_selectors\": [\n      \"[data-test='username'][value='user@bank.com']\",\n      \"[aria-label='Username'][value='user@bank.com']\"\n    ]\n  }\n}\n```\n**ORIGINAL SCRIPT**:\n```python\nimport re\n\n\n    browser = playwright.chromium.launch(headless=False)\n    context = browser.new_context()\n    page = context.new_page()\n    page.goto(\"https://www.saucedemo.com/\")\n    page.locator(\"[data-test=\\\"username\\\"]\").click()\n    page.locator(\"[data-test=\\\"username\\\"]\").fill(\"standard_user\")\n    page.locator(\"[data-test=\\\"password\\\"]\").click()\n    page.locator(\"[data-test=\\\"password\\\"]\").fill(\"secret_sauce\")\n    page.locator(\"[data-test=\\\"login-button\\\"]\").click()\n    page.get_by_role(\"button\", name=\"Open Menu\").click()\n    page.locator(\"[data-test=\\\"logout-sidebar-link\\\"]\").click()\n    page.close()\n\n\n\n    run(playwright)\n```\n\n**PARSED ACTIONS INPUT**:\n```json\n[\n  {\n    \"action_id\": 1,\n    \"action_type\": \"goto\",\n    \"selector\": null,\n    \"value\": \"https://www.saucedemo.com/\",\n    \"description\": \"Navigate to https://www.saucedemo.com/\"\n  },\n  {\n    \"action_id\": 2,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"username\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"username\\\"]\"\n  },\n  {\n    \"action_id\": 3,\n    \"action_type\": \"fill\",\n    \"selector\": \"[data-test=\\\"username\\\"]\",\n    \"value\": \"standard_user\",\n    \"description\": \"Fill the element with selector [data-test=\\\"username\\\"] with 'standard_user'\"\n  },\n  {\n    \"action_id\": 4,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"password\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"password\\\"]\"\n  },\n  {\n    \"action_id\": 5,\n    \"action_type\": \"fill\",\n    \"selector\": \"[data-test=\\\"password\\\"]\",\n    \"value\": \"secret_sauce\",\n    \"description\": \"Fill the element with selector [data-test=\\\"password\\\"] with 'secret_sauce'\"\n  },\n  {\n    \"action_id\": 6,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"login-button\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"login-button\\\"]\"\n  },\n  {\n    \"action_id\": 7,\n    \"action_type\": \"click\",\n    \"selector\": \"get_by_role(\\\"button\\\", name=\\\"Open Menu\\\")\",\n    \"value\": null,\n    \"description\": \"Click on the button with role 'button' and name 'Open Menu'\"\n  },\n  {\n    \"action_id\": 8,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"logout-sidebar-link\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"logout-sidebar-link\\\"]\"\n  }\n]\n```\n\n**OUTPUT REQUIREMENTS**:\n- Keep original action_id, action_type, value, description\n- Replace single \"selector\" with \"primary_selector\" + \"backup_selectors\" array\n- Add wait_conditions for timing-sensitive actions (click, fill, navigation)\n- Add validation_after for state verification when applicable\n- Use selector priority order consistently\n- Return as JSON array with enhanced actions\n\nPlease execute the script to validate selectors and generate enhanced user_actions.json following this exact format.\n\nHuman: \n**TASK**: Generate enhanced user_actions.json with comprehensive selector strategies and testing capabilities.\n\n**CRITICAL REQUIREMENTS**:\n- Take parsed_actions.json as base structure\n- Enhance EACH action with comprehensive selector strategies\n- Add wait conditions and validation for timing-sensitive actions\n- EXECUTE THE COMPLETE SCRIPT to validate selector accuracy\n\n**SELECTOR PRIORITY ORDER** (Use this exact order for ALL selector arrays):\n1. data-test attributes (most reliable)\n2. accessibility selectors ([aria-label], [role], [aria-*])\n3. ID selectors (#element-id)\n4. class selectors (.class-name)\n5. xpath expressions (//element[@attribute='value'])\n6. text-based selectors (element:has-text('Text'))\n\n**ENHANCED ACTION SCHEMA**:\n```json\n{\n  \"action_id\": 1,\n  \"action_type\": \"click\",\n  \"primary_selector\": \"[data-test='username']\",\n  \"backup_selectors\": [\n    \"[aria-label='Username']\",\n    \"#username-field\",\n    \".username-input\",\n    \"//input[@data-test='username']\",\n    \"input[placeholder='Username']\"\n  ],\n  \"value\": \"user@bank.com\",\n  \"description\": \"Fill the username input field\",\n  \"wait_conditions\": {\n    \"wait_type\": \"element_clickable\",\n    \"timeout_ms\": 5000,\n    \"wait_selectors\": [\n      \"[data-test='username']\",\n      \"[aria-label='Username']\",\n      \"#username-field\"\n    ]\n  },\n  \"validation_after\": {\n    \"expected_state\": \"field_filled\",\n    \"validation_selectors\": [\n      \"[data-test='username'][value='user@bank.com']\",\n      \"[aria-label='Username'][value='user@bank.com']\"\n    ]\n  }\n}\n```\n**ORIGINAL SCRIPT**:\n```python\nimport re\n\n\n    browser = playwright.chromium.launch(headless=False)\n    context = browser.new_context()\n    page = context.new_page()\n    page.goto(\"https://www.saucedemo.com/\")\n    page.locator(\"[data-test=\\\"username\\\"]\").click()\n    page.locator(\"[data-test=\\\"username\\\"]\").fill(\"standard_user\")\n    page.locator(\"[data-test=\\\"password\\\"]\").click()\n    page.locator(\"[data-test=\\\"password\\\"]\").fill(\"secret_sauce\")\n    page.locator(\"[data-test=\\\"login-button\\\"]\").click()\n    page.get_by_role(\"button\", name=\"Open Menu\").click()\n    page.locator(\"[data-test=\\\"logout-sidebar-link\\\"]\").click()\n    page.close()\n\n\n\n    run(playwright)\n```\n\n**PARSED ACTIONS INPUT**:\n```json\n[\n  {\n    \"action_id\": 1,\n    \"action_type\": \"goto\",\n    \"selector\": null,\n    \"value\": \"https://www.saucedemo.com/\",\n    \"description\": \"Navigate to https://www.saucedemo.com/\"\n  },\n  {\n    \"action_id\": 2,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"username\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"username\\\"]\"\n  },\n  {\n    \"action_id\": 3,\n    \"action_type\": \"fill\",\n    \"selector\": \"[data-test=\\\"username\\\"]\",\n    \"value\": \"standard_user\",\n    \"description\": \"Fill the element with selector [data-test=\\\"username\\\"] with 'standard_user'\"\n  },\n  {\n    \"action_id\": 4,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"password\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"password\\\"]\"\n  },\n  {\n    \"action_id\": 5,\n    \"action_type\": \"fill\",\n    \"selector\": \"[data-test=\\\"password\\\"]\",\n    \"value\": \"secret_sauce\",\n    \"description\": \"Fill the element with selector [data-test=\\\"password\\\"] with 'secret_sauce'\"\n  },\n  {\n    \"action_id\": 6,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"login-button\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"login-button\\\"]\"\n  },\n  {\n    \"action_id\": 7,\n    \"action_type\": \"click\",\n    \"selector\": \"get_by_role(\\\"button\\\", name=\\\"Open Menu\\\")\",\n    \"value\": null,\n    \"description\": \"Click on the button with role 'button' and name 'Open Menu'\"\n  },\n  {\n    \"action_id\": 8,\n    \"action_type\": \"click\",\n    \"selector\": \"[data-test=\\\"logout-sidebar-link\\\"]\",\n    \"value\": null,\n    \"description\": \"Click on the element with selector [data-test=\\\"logout-sidebar-link\\\"]\"\n  }\n]\n```\n\n**OUTPUT REQUIREMENTS**:\n- Keep original action_id, action_type, value, description\n- Replace single \"selector\" with \"primary_selector\" + \"backup_selectors\" array\n- Add wait_conditions for timing-sensitive actions (click, fill, navigation)\n- Add validation_after for state verification when applicable\n- Use selector priority order consistently\n- Return as JSON array with enhanced actions\n\nPlease execute the script to validate selectors and generate enhanced user_actions.json following this exact format.\n\nAI: {'name': 'browser_navigate', 'arguments': '{\"url\": \"https://www.saucedemo.com/\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.goto('https://www.saucedemo.com/');\n```\n\n### New console messages\n- [VERBOSE] [DOM] Input elements should have autocomplete attributes (suggested: \"current-password\"): ...\n\n### Page state\n- Page URL: https://www.saucedemo.com/\n- Page Title: Swag Labs\n- Page Snapshot:\n```yaml\n- generic [ref=e3]:\n  - generic [ref=e4]: Swag Labs\n  - generic [ref=e5]:\n    - generic [ref=e9]:\n      - textbox \"Username\" [ref=e11]\n      - textbox \"Password\" [ref=e13]\n      - button \"Login\" [ref=e15] [cursor=pointer]\n    - generic [ref=e17]:\n      - generic [ref=e18]:\n        - heading \"Accepted usernames are:\" [level=4] [ref=e19]\n        - text: standard_user\n        - text: locked_out_user\n        - text: problem_user\n        - text: performance_glitch_user\n        - text: error_user\n        - text: visual_user\n      - generic [ref=e20]:\n        - heading \"Password for all users:\" [level=4] [ref=e21]\n        - text: secret_sauce\n```\n\nAI: I have navigated to the login page. I will now execute the login sequence and then generate the enhanced JSON output.{'name': 'browser_click', 'arguments': '{\"element\": \"textbox \\\\\"Username\\\\\"\", \"ref\": \"e11\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.locator('[data-test=\"username\"]').click();\n```\n\n### New console messages\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n- [LOG] Content is cached for offline use. @ https://www.saucedemo.com/static/js/2.84a5cbf3.chunk.js:1\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n\n### Page state\n- Page URL: https://www.saucedemo.com/\n- Page Title: Swag Labs\n- Page Snapshot:\n```yaml\n- generic [ref=e3]:\n  - generic [ref=e4]: Swag Labs\n  - generic [ref=e5]:\n    - generic [ref=e9]:\n      - textbox \"Username\" [active] [ref=e11]\n      - textbox \"Password\" [ref=e13]\n      - button \"Login\" [ref=e15] [cursor=pointer]\n    - generic [ref=e17]:\n      - generic [ref=e18]:\n        - heading \"Accepted usernames are:\" [level=4] [ref=e19]\n        - text: standard_user\n        - text: locked_out_user\n        - text: problem_user\n        - text: performance_glitch_user\n        - text: error_user\n        - text: visual_user\n      - generic [ref=e20]:\n        - heading \"Password for all users:\" [level=4] [ref=e21]\n        - text: secret_sauce\n```\n\nAI: {'name': 'browser_type', 'arguments': '{\"ref\": \"e11\", \"text\": \"standard_user\", \"element\": \"textbox \\\\\"Username\\\\\"\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.locator('[data-test=\"username\"]').fill('standard_user');\n```\n\nAI: {'name': 'browser_click', 'arguments': '{\"ref\": \"e13\", \"element\": \"textbox \\\\\"Password\\\\\"\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.locator('[data-test=\"password\"]').click();\n```\n\n### Page state\n- Page URL: https://www.saucedemo.com/\n- Page Title: Swag Labs\n- Page Snapshot:\n```yaml\n- generic [ref=e3]:\n  - generic [ref=e4]: Swag Labs\n  - generic [ref=e5]:\n    - generic [ref=e9]:\n      - textbox \"Username\" [ref=e11]: standard_user\n      - textbox \"Password\" [active] [ref=e13]\n      - button \"Login\" [ref=e15] [cursor=pointer]\n    - generic [ref=e17]:\n      - generic [ref=e18]:\n        - heading \"Accepted usernames are:\" [level=4] [ref=e19]\n        - text: standard_user\n        - text: locked_out_user\n        - text: problem_user\n        - text: performance_glitch_user\n        - text: error_user\n        - text: visual_user\n      - generic [ref=e20]:\n        - heading \"Password for all users:\" [level=4] [ref=e21]\n        - text: secret_sauce\n```\n\nAI: {'name': 'browser_type', 'arguments': '{\"ref\": \"e13\", \"text\": \"secret_sauce\", \"element\": \"textbox \\\\\"Password\\\\\"\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.locator('[data-test=\"password\"]').fill('secret_sauce');\n```\n\nAI: {'name': 'browser_click', 'arguments': '{\"ref\": \"e15\", \"element\": \"button \\\\\"Login\\\\\"\"}'}\nTool: ### Ran Playwright code\n```js\nawait page.locator('[data-test=\"login-button\"]').click();\n```\n\n### New console messages\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n- [ERROR] Failed to load resource: the server responded with a status of 401 (Unauthorized) @ https://...\n\n### Page state\n- Page URL: https://www.saucedemo.com/inventory.html\n- Page Title: Swag Labs\n- Page Snapshot:\n```yaml\n- generic [ref=e22]:\n  - generic [ref=e23]:\n    - generic [ref=e24]:\n      - generic [ref=e25]:\n        - generic [ref=e26]:\n          - button \"Open Menu\" [ref=e27] [cursor=pointer]\n          - img \"Open Menu\" [ref=e28]\n        - generic [ref=e30]: Swag Labs\n      - generic [ref=e33]:\n        - generic [ref=e34]: Products\n        - combobox [ref=e37]:\n          - option \"Name (A to Z)\" [selected]\n          - option \"Name (Z to A)\"\n          - option \"Price (low to high)\"\n          - option \"Price (high to low)\"\n    - generic [ref=e41]:\n      - generic [ref=e42]:\n        - link \"Sauce Labs Backpack\" [ref=e44] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Sauce Labs Backpack\" [ref=e45] [cursor=pointer]\n        - generic [ref=e46]:\n          - generic [ref=e47]:\n            - link \"Sauce Labs Backpack\" [ref=e48] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e49] [cursor=pointer]: Sauce Labs Backpack\n            - generic [ref=e50]: carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection.\n          - generic [ref=e51]:\n            - generic [ref=e52]: $29.99\n            - button \"Add to cart\" [ref=e53] [cursor=pointer]\n      - generic [ref=e54]:\n        - link \"Sauce Labs Bike Light\" [ref=e56] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Sauce Labs Bike Light\" [ref=e57] [cursor=pointer]\n        - generic [ref=e58]:\n          - generic [ref=e59]:\n            - link \"Sauce Labs Bike Light\" [ref=e60] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e61] [cursor=pointer]: Sauce Labs Bike Light\n            - generic [ref=e62]: A red light isn't the desired state in testing but it sure helps when riding your bike at night. Water-resistant with 3 lighting modes, 1 AAA battery included.\n          - generic [ref=e63]:\n            - generic [ref=e64]: $9.99\n            - button \"Add to cart\" [ref=e65] [cursor=pointer]\n      - generic [ref=e66]:\n        - link \"Sauce Labs Bolt T-Shirt\" [ref=e68] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Sauce Labs Bolt T-Shirt\" [ref=e69] [cursor=pointer]\n        - generic [ref=e70]:\n          - generic [ref=e71]:\n            - link \"Sauce Labs Bolt T-Shirt\" [ref=e72] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e73] [cursor=pointer]: Sauce Labs Bolt T-Shirt\n            - generic [ref=e74]: Get your testing superhero on with the Sauce Labs bolt T-shirt. From American Apparel, 100% ringspun combed cotton, heather gray with red bolt.\n          - generic [ref=e75]:\n            - generic [ref=e76]: $15.99\n            - button \"Add to cart\" [ref=e77] [cursor=pointer]\n      - generic [ref=e78]:\n        - link \"Sauce Labs Fleece Jacket\" [ref=e80] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Sauce Labs Fleece Jacket\" [ref=e81] [cursor=pointer]\n        - generic [ref=e82]:\n          - generic [ref=e83]:\n            - link \"Sauce Labs Fleece Jacket\" [ref=e84] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e85] [cursor=pointer]: Sauce Labs Fleece Jacket\n            - generic [ref=e86]: It's not every day that you come across a midweight quarter-zip fleece jacket capable of handling everything from a relaxing day outdoors to a busy day at the office.\n          - generic [ref=e87]:\n            - generic [ref=e88]: $49.99\n            - button \"Add to cart\" [ref=e89] [cursor=pointer]\n      - generic [ref=e90]:\n        - link \"Sauce Labs Onesie\" [ref=e92] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Sauce Labs Onesie\" [ref=e93] [cursor=pointer]\n        - generic [ref=e94]:\n          - generic [ref=e95]:\n            - link \"Sauce Labs Onesie\" [ref=e96] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e97] [cursor=pointer]: Sauce Labs Onesie\n            - generic [ref=e98]: Rib snap infant onesie for the junior automation engineer in development. Reinforced 3-snap bottom closure, two-needle hemmed sleeved and bottom won't unravel.\n          - generic [ref=e99]:\n            - generic [ref=e100]: $7.99\n            - button \"Add to cart\" [ref=e101] [cursor=pointer]\n      - generic [ref=e102]:\n        - link \"Test.allTheThings() T-Shirt (Red)\" [ref=e104] [cursor=pointer]:\n          - /url: \"#\"\n          - img \"Test.allTheThings() T-Shirt (Red)\" [ref=e105] [cursor=pointer]\n        - generic [ref=e106]:\n          - generic [ref=e107]:\n            - link \"Test.allTheThings() T-Shirt (Red)\" [ref=e108] [cursor=pointer]:\n              - /url: \"#\"\n              - generic [ref=e109] [cursor=pointer]: Test.allTheThings() T-Shirt (Red)\n            - generic [ref=e110]: This classic Sauce Labs t-shirt is perfect to wear when cozying up to your keyboard to automate a few tests. Super-soft and comfy ringspun combed cotton.\n          - generic [ref=e111]:\n            - generic [ref=e112]: $15.99\n            - button \"Add to cart\" [ref=e113] [cursor=pointer]\n  - contentinfo [ref=e114]:\n    - list [ref=e115]:\n      - listitem [ref=e116]:\n        - link \"Twitter\" [ref=e117] [cursor=pointer]:\n          - /url: https://twitter.com/saucelabs\n      - listitem [ref=e118]:\n        - link \"Facebook\" [ref=e119] [cursor=pointer]:\n          - /url: https://www.facebook.com/saucelabs\n      - listitem [ref=e120]:\n        - link \"LinkedIn\" [ref=e121] [cursor=pointer]:\n          - /url: https://www.linkedin.com/company/sauce-labs/\n    - generic [ref=e122]: © 2025 Sauce Labs. All Rights Reserved. Terms of Service | Privacy Policy\n```"
  ]
}
[llm/end] [tool:user_actions_enhancer > chain:RunnableSequence > llm:ChatGoogleGenerativeAI] [12.61s] Exiting LLM run with output:
{
  "generations": [
    [
      {
        "text": "",
        "generation_info": {
          "finish_reason": "STOP",
          "model_name": "gemini-2.5-pro",
          "safety_ratings": []
        },
        "type": "ChatGenerationChunk",
        "message": {
          "lc": 1,
          "type": "constructor",
          "id": [
            "langchain",
            "schema",
            "messages",
            "AIMessageChunk"
          ],
          "kwargs": {
            "content": "",
            "response_metadata": {
              "finish_reason": "STOP",
              "model_name": "gemini-2.5-pro",
              "safety_ratings": []
            },
            "type": "AIMessageChunk",
            "id": "run--25eb0043-398b-4b04-baf5-f05e818ff92f",
            "usage_metadata": {
              "input_tokens": 8377,
              "output_tokens": 3,
              "total_tokens": 8380,
              "input_token_details": {
                "cache_read": 0
              },
              "output_token_details": {
                "reasoning": 3
              }
            },
            "tool_calls": [],
            "invalid_tool_calls": []
          }
        }
      }
    ]
  ],
  "llm_output": null,
  "run": null,
  "type": "LLMResult"
}
[chain/start] [tool:user_actions_enhancer > chain:RunnableSequence > parser:ToolsAgentOutputParser] Entering Parser run with input:
[inputs]
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence > parser:ToolsAgentOutputParser] [1ms] Exiting Parser run with output:
[outputs]
[chain/end] [tool:user_actions_enhancer > chain:RunnableSequence] [12.62s] Exiting Chain run with output:
[outputs]
2025-08-12 11:09:12,303 - mcp_use - INFO - ✅ Agent finished at step 7
2025-08-12 11:09:12 - mcp_use - INFO - ✅ Agent finished at step 7
2025-08-12 11:09:12,304 - mcp_use - WARNING - ⚠️ Agent stopped after reaching max iterations (1000)
2025-08-12 11:09:12 - mcp_use - WARNING - ⚠️ Agent stopped after reaching max iterations (1000)
2025-08-12 11:09:12,305 - mcp_use - INFO - 🎉 Agent execution complete in 60.60093808174133 seconds
2025-08-12 11:09:12 - mcp_use - INFO - 🎉 Agent execution complete in 60.60093808174133 seconds
2025-08-12 11:09:12 - engines.DiscoveryEngine.Tools.PageJourneyTool - ERROR - ❌ User actions output validation failed: User actions parsing failed: Expecting value: line 1 column 1 (char 0)
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE 
[tool/error] [tool:user_actions_enhancer] [61.29s] Tool run errored with error:
OutputParserException('UserActions schema validation failed: User actions parsing failed: Expecting value: line 1 column 1 (char 0)\nFor troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE \nFor troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE ')Traceback (most recent call last):


  File "/Users/ramacharanreddykasireddy/ReQon/reqon-mcp-service/engines/DiscoveryEngine/Tools/PageJourneyTool.py", line 625, in enhance_user_actions_tool
    raise OutputParserException(f"User actions parsing failed: {parsed_data['parse_error']}")


langchain_core.exceptions.OutputParserException: User actions parsing failed: Expecting value: line 1 column 1 (char 0)
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE 



During handling of the above exception, another exception occurred:



Traceback (most recent call last):


  File "/Users/ramacharanreddykasireddy/ReQon/reqon-mcp-service/.venv/lib/python3.11/site-packages/langchain_core/tools/base.py", line 970, in arun
    response = await coro_with_context(coro, context)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/Users/ramacharanreddykasireddy/ReQon/reqon-mcp-service/.venv/lib/python3.11/site-packages/langchain_core/tools/structured.py", line 110, in _arun
    return await self.coroutine(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/Users/ramacharanreddykasireddy/ReQon/reqon-mcp-service/engines/DiscoveryEngine/Tools/PageJourneyTool.py", line 639, in enhance_user_actions_tool
    raise OutputParserException(f"UserActions schema validation failed: {str(e)}")


langchain_core.exceptions.OutputParserException: UserActions schema validation failed: User actions parsing failed: Expecting value: line 1 column 1 (char 0)
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE 
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE
2025-08-12 11:09:12 - engines.DiscoveryEngine.Tools.PageJourneyTool - ERROR - ❌ User actions LangChain validation failed: UserActions schema validation failed: User actions parsing failed: Expecting value: line 1 column 1 (char 0)
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE 
For troubleshooting, visit: https://python.langchain.com/docs/troubleshooting/errors/OUTPUT_PARSING_FAILURE 
2025-08-12 11:09:12 - app.utils.retry_manager - WARNING - ❌ RetryManager: Attempt 1 failed after 61.30s - UserActions schema validation failed: User actions parsing failed: Expecting value: line 1 column 1 (char 0)
For troubleshooting, visit: https://pytho

now I got this debug information "llm_output": null, then we are seeing the error ⚠️ Agent stopped after reaching max iterations (1000)

--- a/mcp_use/agents/mcpagent.py
+++ b/mcp_use/agents/mcpagent.py
@@ -492,6 +492,10 @@ class MCPAgent:
 
             logger.info(f"🏁 Starting agent execution with max_steps={steps}")
 
+            # Track whether agent finished successfully vs reached max iterations
+            agent_finished_successfully = False
+            result = None
+
             # Create a run manager with our callbacks if we have any - ONCE for the entire execution
             run_manager = None
             if self.callbacks:
@@ -578,6 +582,7 @@ class MCPAgent:
                     # Process the output
                     if isinstance(next_step_output, AgentFinish):
                         logger.info(f"✅ Agent finished at step {step_num + 1}")
+                        agent_finished_successfully = True
                         result = next_step_output.return_values.get("output", "No output generated")
                         # End the chain if we have a run manager
                         if run_manager:
@@ -659,6 +664,7 @@ class MCPAgent:
                         tool_return = self._agent_executor._get_tool_return(last_step)
                         if tool_return is not None:
                             logger.info(f"🏆 Tool returned directly at step {step_num + 1}")
+                            agent_finished_successfully = True
                             result = tool_return.return_values.get("output", "No output generated")
                             break
 
@@ -681,10 +687,16 @@ class MCPAgent:
 
             # --- Loop finished ---
             if not result:
-                logger.warning(f"⚠️ Agent stopped after reaching max iterations ({steps})")
-                result = f"Agent stopped after reaching the maximum number of steps ({steps})."
-                if run_manager:
-                    await run_manager.on_chain_end({"output": result})
+                if agent_finished_successfully:
+                    # Agent finished successfully but returned empty output
+                    result = "Agent completed the task successfully."
+                    logger.info("✅ Agent finished successfully with empty output")
+                else:
+                    # Agent actually reached max iterations
+                    logger.warning(f"⚠️ Agent stopped after reaching max iterations ({steps})")
+                    result = f"Agent stopped after reaching the maximum number of steps ({steps})."
+                    if run_manager:
+                        await run_manager.on_chain_end({"output": result})
 
             # If structured output was requested but not achieved, attempt one final time
             if output_schema and structured_llm and not success:

